generation/src/proof_generation/interpreter/hilbert_interpreter.py

```
@@ -0,0 +1,310 @@
+        if self.prop == 1:
+            self.complete_delta(delta, 2)
+            self.hproof.add_proof_step(HProp1(self.index, ret.conclusion, delta[0], delta[1]))
+        elif self.prop == 2:
+            self.complete_delta(delta, 3)
+            self.hproof.add_proof_step(HProp2(self.index, ret.conclusion, delta[0], delta[1], delta[2]))
+        elif self.prop == 3:
+            self.complete_delta(delta, 2)
+            self.hproof.add_proof_step(HProp3(self.index, ret.conclusion, delta[0], delta[1]))
+        self.prop = 0
+        self.steps.setdefault(str(ret.conclusion), self.index)
+        for value in delta.values():
+            key = str(value)
+            argument = self.arguments.get(key)
+            if argument is None:
+                self.argument_index = self.argument_index + 1
+                argument = HArgument(self.argument_index, value)
+                self.arguments[key] = argument
+            argument.increment_count()
+        return ret
+
+    def exists_quantifier(self) -> Proved:
+        ret = super().exists_quantifier()
+        self.index = self.index + 1
+        self.hproof.add_proof_step(HExistsQuantifier(self.index, ret.conclusion))
+        self.steps.setdefault(str(ret.conclusion), self.index)
+        return ret
+
+    def exists_generalization(self, proved: Proved, var: EVar) -> Proved:
+        ret = super().exists_generalization(proved, var)
+        self.index = self.index + 1
+        idx = self.steps[str(proved.conclusion)]
+        self.hproof.add_proof_step(HExistsGeneralization(self.index, ret.conclusion, proved.conclusion, var, idx))
+        self.steps.setdefault(str(ret.conclusion), self.index)
+        return ret
+
+    def publish_proof(self, proved: Proved) -> None:
+        super().publish_proof(proved)
+        self.hproof.set_arguments(list(self.arguments.values()))
+
+        self.out.write(f'{str(self.hproof)}\n')
+
+    def publish_axiom(self, axiom: Pattern) -> None:
```

This handling of axioms feels sketchy. If we don't handle any Hilbert proofs using axioms yet, let's raise an Error here and note to consider it more carefully once we do. 

The thing is that `publish_axiom` and `publish_claim` are executed in the Gamma/Claim phases and those aren't really proofs, but "constructions". Now you do indeed need some way to get the axioms you'll be using in the Proof phase inside the interpreter, but it's not clear to me that _this_ interpreter will always (or even should) run the Gamma phase beforehand. It makes sense to grab them (via constructor) from the `ProofExp` class instead I think?  

@nishantjr please comment on this.

---

I've added code to provide more intelligible messages for errors which would arise from the Axioms phase not being called here: https://github.com/Pi-Squared-Network/pi2/blob/74ed4295cced4ca05bb0fc5ca229d74aa4a9fd51/generation/src/proof_generation/interpreter/hilbert_interpreter.py#L193